dev/com.ibm.ws.session.cache/src/com/ibm/ws/session/store/cache/CacheHashMapMR.java

```diff
@@ -152,67 +157,65 @@ protected Object getValue(String id, BackedSession s) {
             pmiStats.readTimes(sessionPropBytes == null ? 0 : sessionPropBytes.length, System.currentTimeMillis() - startTime);
         }
 
+        // Before returning the value, confirm that the cache entry for the session hasn't expired
+        if (!cacheStoreService.cache.containsKey(createSessionKey(sessId, appName))) {
+            if (trace && tc.isDebugEnabled())
+                Tr.debug(this, tc, sessId + " does not appear to be a valid session for " + appName);
+            tmp = null;
+            throw new UnsupportedOperationException(); // TODO implement code path where cache entry for session is expired. Delete the property entry?
+        }
+
         return tmp;
     }
 
     /**
      * copied from DatabaseHashMapMR
      */
     @Override
-    @SuppressWarnings({ "rawtypes", "unused" })
+    @SuppressWarnings({ "rawtypes", "unchecked" })
     boolean handlePropertyHits(BackedSession d2) {
         final boolean trace = TraceComponent.isAnyTracingEnabled();
 
         Thread t = Thread.currentThread();
 
-        Enumeration vEnum = null;
-        boolean doWrite = false;
-
         String id = d2.getId();
-        String propid = null;
+        String appName = getAppName();
 
         try {
+            Set<String> propsToWrite = null;
+
             // we are not synchronized here - were not in old code either
             Hashtable sht = null;
             if (_smc.writeAllProperties()) {
                 Hashtable ht = d2.getSwappableData();
-                vEnum = ht.keys();
-                doWrite = true;
+                propsToWrite = ht.keySet();
                 if (trace && tc.isDebugEnabled()) {
                     Tr.debug(this, tc, "doing app changes for ALL mSwappable Data", ht);
                 }
             } else {
                 if (d2.appDataChanges != null) {
                     if (appDataTablesPerThread) {
                         if ((sht = (Hashtable) d2.appDataChanges.get(t)) != null) {
-                            doWrite = true;
                             if (trace && tc.isDebugEnabled()) {
                                 Tr.debug(this, tc, "doing app changes for " + id + " on thread " + t);
                             }
-                            if (sht != null) {
-                                vEnum = sht.keys();
-                            } else {
-                                vEnum = (new Hashtable()).keys();
-                            }
+                            propsToWrite = sht.keySet();
```

Comment:
`sht` is an unfortunate variable name...  Change to `syncHt`?

Reply:
I don't like the name either.  Currently it matches DatabaseHashMapMR from which this code was copied.  At this point, the code has diverged enough it should be okay to change it